src/java_bytecode/ci_lazy_methods.cpp

```diff
@@ -1,39 +1,483 @@
+/// type *and* a constructor callsite indicating an object of that type may be
+/// instantiated (or evidence that an object of that type exists before the main
+/// function is entered, such as being passed as a parameter).
+/// Elaborates lazily-converted methods that may be reachable starting
+/// from the main entry point (usually provided with the --function command-
+/// line option
+/// \param symbol_table: global symbol table
+/// \param [out] lazy_methods: map from method names to relevant symbol and
+///   parsed-method objects.
+/// \return Returns false on success
+bool ci_lazy_methodst::operator()(
+  symbol_tablet &symbol_table,
+  lazy_methodst &lazy_methods,
+  std::function<void(
+    const symbolt &,
+    const java_bytecode_parse_treet::methodt &,
+    ci_lazy_methods_neededt)> method_converter)
+{
+  std::vector<irep_idt> method_worklist1;
+  std::vector<irep_idt> method_worklist2;
+
+  auto main_function=
+    get_main_symbol(symbol_table, main_class, get_message_handler(), true);
+  if(main_function.stop_convert)
+  {
+    // Failed, mark all functions in the given main class(es) reaclass_hierarchyable.
+    std::vector<irep_idt> reaclass_hierarchyable_classes;
+    if(!main_class.empty())
+      reaclass_hierarchyable_classes.push_back(main_class);
+    else
+      reaclass_hierarchyable_classes=main_jar_classes;
+    for(const auto &classname : reaclass_hierarchyable_classes)
+    {
+      const auto &methods=
+        java_class_loader.class_map.at(classname).parsed_class.methods;
+      for(const auto &method : methods)
+      {
+        const irep_idt methodid="java::"+id2string(classname)+"."+
+          id2string(method.name)+":"+
+          id2string(method.signature);
+        method_worklist2.push_back(methodid);
+      }
+    }
+  }
+  else
+    method_worklist2.push_back(main_function.main_function.name);
+
+  // Add any extra entry points specified; we should elaborate these in the
+  // same way as the main function.
+  std::vector<irep_idt> extra_entry_points=lazy_methods_extra_entry_points;
+  resolve_method_names(extra_entry_points, symbol_table);
+  method_worklist2.insert(
+    method_worklist2.begin(),
+    extra_entry_points.begin(),
+    extra_entry_points.end());
+
+  std::set<irep_idt> needed_classes;
+
+  {
+    std::vector<irep_idt> needed_clinits;
+    ci_lazy_methods_neededt initial_lazy_methods(
+      needed_clinits,
+      needed_classes,
+      symbol_table);
+    initialize_needed_classes(
+      method_worklist2,
+      namespacet(symbol_table),
+      initial_lazy_methods);
+    method_worklist2.insert(
+      method_worklist2.end(),
+      needed_clinits.begin(),
+      needed_clinits.end());
+  }
+
+  std::set<irep_idt> methods_already_populated;
+  std::vector<const code_function_callt *> virtual_callsites;
 
-#include <string>
+  bool any_new_methods;
+  do
+  {
+    any_new_methods=false;
+    while(!method_worklist2.empty())
```

Comment:
I'm not a fan of this nested loop. Could we refactor this loop body into a method which returns whether there are any new methods?